thoth/storages/graph/postgres.py

```diff
@@ -4671,6 +4671,80 @@ def sync_dependency_monkey_result(self, document: dict) -> None:
 
                 inspection_run.dependency_monkey_run_id = dependency_monkey_run.id
 
+    def get_python_package_required_symbols(
+        self, package_name: str, package_version: str, index_url: str
+    ) -> List[str]:
+        """Get required symbols for a Python package in a specified version."""
+        package_name = self.normalize_python_package_name(package_name)
+        package_version = self.normalize_python_package_version(package_version)
+
+        with self._session_scope() as session:
+            query = (
+                session.query(PythonPackageVersionEntity)
+                .filter(PythonPackageVersionEntity.package_name == package_name)
+                .filter(PythonPackageVersionEntity.package_version == package_version)
+                .join(
+                    PythonPackageIndex,
+                    PythonPackageIndex.id == PythonPackageVersionEntity.python_package_index_id
+                )
+                .filter(PythonPackageIndex.url == index_url)
+                .join(
+                    HasArtifact,
+                    PythonPackageVersionEntity.id == HasArtifact.python_package_version_entity_id
+                )
+                .join(
+                    RequiresSymbol,
+                    RequiresSymbol.python_artifact_id == HasArtifact.python_artifact_id
+                )
+                .join(
+                    VersionedSymbol,
+                    RequiresSymbol.versioned_symbol_id == VersionedSymbol.id
+                )
+                .with_entities(VersionedSymbol.symbol)
+            )
+
+            # Query returns list of single tuples
+            # NOTE: can be empty even if request is valid
+            to_return = [i[0] for i in query.all()]
+            _LOGGER.debug("The package requires the following symbols: %r", str(to_return))
+            return [i[0] for i in query.all()]
+
+    def get_analyzed_image_symbols_all(
+        self,
+        os_name: str,
+        os_version: str,
+        *,
+        python_version: Optional[str] = None,
+        cuda_version: Optional[str] = None,
+    ) -> List[str]:
+        """Get symbols associated with a given image."""
+        with self._session_scope() as session:
+            query = (
+                session.query(PackageExtractRun)
+                .filter(PackageExtractRun.os_id == os_name)
+                .filter(PackageExtractRun.os_version_id == os_version)
+                .join(
+                    ExternalSoftwareEnvironment,
+                    PackageExtractRun.external_software_environment_id == ExternalSoftwareEnvironment.id
+                )
+                .filter(ExternalSoftwareEnvironment.cuda_version == cuda_version)
+                .filter(ExternalSoftwareEnvironment.python_version == python_version)
```

Comment:
It will be good to add these filters conditionally I think. Otherwise, you will be matching to `None` values in tables if `python_version` and `cuda_version` will not be provided. (I guess that's not expected behaviour?)

Reply:
That was what I planned, if they are not provided I only wanted to match with rows that had `NULL` in those columns.  Otherwise, we might match multiple `cuda_versions` if we don't filter on it.

Reply:
OK, that makes sense :+1: 

BTW please use `SoftwareEnvironment` - that's software environment we suggest users to use (e.g. our s2i-thoth container images). `ExternalSoftwareEnvironment` can be any environment that a user submitted for analysis and we do not have any control about it.